linalg/hypre.cpp

```diff
@@ -1616,7 +1819,16 @@ HypreParMatrix * RAP(const HypreParMatrix *A, const HypreParMatrix *P)
       hypre_ParCSRMatrixOwnsColStarts((hypre_ParCSRMatrix*)(*P));
 
    hypre_ParCSRMatrix * rap;
+#ifdef HYPRE_USING_CUDA
+   {
+      hypre_ParCSRMatrix *Q = hypre_ParCSRMatMat(*A,*P);
+      const bool keepTranspose = false;
+      rap = hypre_ParCSRTMatMatKT(*P,Q,keepTranspose);
+      delete Q;
```

Comment:
@v-dobrev valgrind complains about a mismatched `free() / delete / delete[]`. It does not say exactly where, but it looks like the `delete_hypre*` functions should be called rather than `delete Q`. Is the best/usual way to do this by using the constructor `HypreParMatrix(hypre_ParCSRMatrix *a, true)`, and then relying on `HypreParMatrix::Destroy()` called by `~HypreParMatrix`?

Reply:
Since this object is created by hypre, we need to destroy it by calling a hypre function: `hypre_ParCSRMatrixDestroy`.